### PR TITLE
Add question-answering model

### PR DESCRIPTION
**Pull request by kamalkraj**
Model conversion works perfectly as specified in the readme.

### CODE REVIEW

**Comment by pommedeterresautee** [comment]
Hi @kamalkraj, thank you a lot for the work!!! It's huge and would make the lib kind of "feature complete" at least regarding the supported tasks. The `notebook` thing is really a good idea, and probably something we should do for other tasks.

I am wondering if it's possible to reduce the code from HF in the `question_answering.py` file the same way than in #93 for token classification task? It would make this lib more robust to HF updates...

**Comment by kamalkraj** [comment]
Done @pommedeterresautee

**Comment by pommedeterresautee** [comment]
I finally had the time to dig a bit in this very interesting PR.
The code works when run manually, etc.
However, there is an issue when the batch size is set to b > 1:

from tests:

```python
collecting ... collected 1 item

test_models.py::test_qa_gpu FAILED                                       [100%]2022-06-26 20:06:54.832001940 [E:onnxruntime:, sequential_executor.cc:368 Execute] Non-zero status code returned while running EmbedLayerNormalization node. Name:'EmbedLayerNormalization_0' Status Message: input_ids and position_ids shall have same shape

tests/test_models.py:248 (test_qa_gpu)
@pytest.mark.gpu
    def test_qa_gpu():
        commands = [
            "--model",
            "distilbert-base-cased-distilled-squad",
            "--task",
            "question-answering",
            "--backend",
            "onnx",
            "--batch",
            "2",
            "2",
            "2",
            "--seq-len",
            "8",
            "8",
            "8",
            "--output",
            tempfile.mkdtemp(),
        ]
        args = parse_args(commands=commands)
>       main(commands=args)

test_models.py:270: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/geantvert/workspace/fast_transformer/src/transformer_deploy/convert.py:365: in main
    ort_output, time_buffer = launch_inference(
/home/geantvert/workspace/fast_transformer/src/transformer_deploy/convert.py:115: in launch_inference
    output = infer(batch_input)
/home/geantvert/workspace/fast_transformer/src/transformer_deploy/convert.py:361: in infer_ort
    results = inference_onnx_binding(model_onnx=ort_model, inputs=inputs, device=commands.device)
/home/geantvert/workspace/fast_transformer/src/transformer_deploy/backends/ort_utils.py:256: in inference_onnx_binding
    model_onnx.run_with_iobinding(binding)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <onnxruntime.capi.onnxruntime_inference_collection.InferenceSession object at 0x7f47402f44f0>
iobinding = <onnxruntime.capi.onnxruntime_inference_collection.IOBinding object at 0x7f473e6b87c0>
run_options = None

    def run_with_iobinding(self, iobinding, run_options=None):
        """
         Compute the predictions.
    
         :param iobinding: the iobinding object that has graph inputs/outputs bind.
         :param run_options: See :class:`onnxruntime.RunOptions`.
        """
>       self._sess.run_with_iobinding(iobinding._iobinding, run_options)
E       RuntimeError: Error in execution: Non-zero status code returned while running EmbedLayerNormalization node. Name:'EmbedLayerNormalization_0' Status Message: input_ids and position_ids shall have same shape

```

I have tried with onnxruntime 1.11.1 and 1.12.0.
@kamalkraj is it something you are aware of?

**Comment by kamalkraj** [comment]
Hi @pommedeterresautee,

I am not aware of this error. I will look into it.

**Comment by kamalkraj** [comment]
Hi @pommedeterresautee ,

The above issue is particularly related with `distilbert` when run onnx optimization with batch size >= 2. (I have checked with bert-base model with batch size >=2, no issue)
It is not related to this QA model. The same issue can be replicated with classification model also on the main branch.
more details on #106 

We can fix this by passing dummy tensors to onnx model with always batch size 1. 
The `batch size` and `seq_len` parameters can be used for only `TensorRT`.

Let me know your thoughts on it.

**Comment by pommedeterresautee** [comment]
Thank you @kamalkraj for your analysis.
If the source of the issues is the optimization process may be we can check if disabling `enable_embed_layer_norm` in `FusionOptions` for distilbert and open an issue on ORT? (so no more kernel fusion)
In our experience most ORT kernel fusions have little to no effect on latency anyway (at least in batch size > 1 setup).

**Comment by kamalkraj** [comment]
With `enable_embed_layer_norm` disabled model conversion works perfectly with all batch sizes 

```python
optimization_options = FusionOptions(model_type=architecture)
optimization_options.enable_embed_layer_norm = False # Distilbert bug fix
```

@pommedeterresautee

**Comment by kamalkraj** [comment]
Can we have this fix for distilbert ? Until this get fixed in ORT
@pommedeterresautee

**Comment by pommedeterresautee** [comment]
Yeah I think it's ok to set this value to false in this PR, so we can merge it, and open an issue on ORT repo in parallel.

**Comment by kamalkraj** [comment]
@pommedeterresautee 
I am actually trying to replicate the error using HF Transformers and ORT. 
HF Transformers also does the onnx conversion using batch size 2.
But i am unable to replicate the error.

Notebook -> https://colab.research.google.com/drive/1san0_xGSO-97WJOqtgOjn-i-ezsU9fYp?usp=sharing

Please let me know if i am missing something

**Comment by kamalkraj** [comment]
@pommedeterresautee 
I have updated the Notebook 

The problem is with running model inference onnx_binding.
I have checked without using it and optimization works without disabling `enable_embed_layer_norm`

```python
from typing import Union

import numpy as np
from onnxruntime import InferenceSession
from transformers import AutoConfig, AutoTokenizer, PreTrainedModel, QuestionAnsweringPipeline
from transformers.modeling_outputs import QuestionAnsweringModelOutput


# to suppress a warning, use a known class name
# https://github.com/huggingface/transformers/blob/3f936df66287f557c6528912a9a68d7850913b9b/
# src/transformers/pipelines/base.py#L882
class BertForQuestionAnswering(PreTrainedModel):
    def __init__(self, model_name: str, model_path: str):
        config = AutoConfig.from_pretrained(model_name)
        super().__init__(config)
        self.model_name = model_name
        self.model_path = model_path
        self.session = InferenceSession(
            f"triton_models/{model_path}", providers=["CUDAExecutionProvider"]
        )

    def forward(
        self,
        input_ids=None,
        attention_mask=None,
        token_type_ids: Union[None, np.int32] = None,
    ):
        inputs = {
            "input_ids": np.int32(input_ids.numpy()),
            "attention_mask": np.int32(attention_mask.numpy()),
        }
        # print(inputs)
        if token_type_ids is not None:
            inputs["token_type_ids"] = np.int32(token_type_ids.numpy())

        outputs = self.session.run(
            output_names=["start_logits", "end_logits"], input_feed=dict(inputs)
        )
        # print(outputs)
        return QuestionAnsweringModelOutput(
            start_logits=outputs[0], end_logits=outputs[1]
        )


current_name = "distilbert-base-cased-distilled-squad"

model = BertForQuestionAnswering(
    model_name=current_name, model_path="model-original.onnx"
)
tokenizer = AutoTokenizer.from_pretrained(current_name)

onnx_qa = QuestionAnsweringPipeline(model=model, tokenizer=tokenizer)

question = "Where do I live?"
context = "My name is Wolfgang and I live in Berlin"
pred = onnx_qa(question, context)

print(pred)


model = BertForQuestionAnswering(model_name=current_name, model_path="model.onnx")
onnx_qa = QuestionAnsweringPipeline(model=model, tokenizer=tokenizer)

pred = onnx_qa(question, context)
print(pred)

```

**Comment by pommedeterresautee** [comment]
May I ask you to try with the branch `t5_3b` ? The inference part has been rewrote to be a bit more robust.

**Comment by kamalkraj** [comment]
The same error with `t5_3b` branch also.

The actual issue is with onnx optimization , i have successfully reproduced the error and opened issue https://github.com/microsoft/onnxruntime/issues/12064

I will disable the  `enable_embed_layer_norm` and push the fix for now.

**Comment by kamalkraj** [comment]
disabled `enable_embed_layer_norm` https://github.com/ELS-RD/transformer-deploy/pull/84/commits/bc568fc7db89b7b0381701cef9bad5d7e4730f73

**Review by pommedeterresautee — COMMENTED** [review]
Still need a small change in the architecture check to have all tests working (including GPU one).

File: src/transformer_deploy/backends/ort_utils.py [hunk 1]
```diff
@@ -103,10 +103,13 @@ def optimize_onnx(
     :param hidden_size: hidden layer size of a model (0 -> try to detect)
     :param architecture: model architecture to optimize. One of [bert, bart, gpt2]
     """
-    assert architecture in ["bert", "bart", "gpt2"], f"unsupported architecture: {architecture}"
-    opt_level = 1 if architecture == "bert" else 0
+    assert architecture in ["bert", "bart", "distilbert", "gpt2"], f"unsupported architecture: {architecture}"
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply src/transformer_deploy/backends/ort_utils.py]: Done https://github.com/ELS-RD/transformer-deploy/pull/84/commits/dd8cfae16f300de87a3b9752493ea6ab408b01e2

**Pull request merged**